### PR TITLE
Cars homework

### PR DESCRIPTION
**Pull request by niemarcin**
- add `speed_` protected member to Car class
- change `speed_` only if `accelerate()` parameter >= 0
- engines members are private - encapsulation
- engines as members of cars are protected - to be accessible in derived classes
- exceptions for InvalidGear, MovingCar and InvalidEngine
- check for proper engine type in changeEngine
- check if car is not moving on changeEngine call
- check if changing gears is valid
- class members are private or protected - encapsulation
- tests added
- workflow file added

### CODE REVIEW

File: src/Car.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+void Car::brake() {
+    std::cout << __FUNCTION__ << std::endl;
+    speed_ = 0;
+}
+
+void Car::accelerate(int speed) {
+    std::cout << __FUNCTION__ << std::endl;
+    speed_ = speed > 0 ? speed : 0;
```
Comment (WBigaj) [inline src/Car.cpp:29]: Why you do not throw error when speed is negative?

File: src/ElectricCar.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+void ElectricCar::charge() {
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void ElectricCar::restore() {
+    std::cout << __FUNCTION__ << std::endl;
+    charge();
+}
+
+void ElectricCar::changeEngine(Engine* engine) {
```
Comment (WBigaj) [inline src/ElectricCar.cpp:26]: Can't you just take ElectricEngine as parameter?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

File: src/Car.cpp (hunk 1 above)
Comment (MaciejSalamonski) [reply src/Car.cpp:29]: It wasn't necessary, but I think it would be nice

File: src/ElectricCar.cpp (hunk 2 above)
Comment (MaciejSalamonski) [reply src/ElectricCar.cpp:26]: In my opinion its good. What if the car is a hybrid? Good thing that we can change different types of engines. The code is more universal. Marcin secured it well. The function throws an exception when the wrong engine type is given.

**Review by MaciejSalamonski — APPROVED** [review]
Very nice code! Have a good day!

File: src/ElectricCar.cpp (hunk 2 above)
Comment (niemarcin) [reply src/ElectricCar.cpp:26]: It's an overridden method from Car class `virtual void changeEngine(Engine* engine) = 0;`

**Review by ziobron — APPROVED** [review]
Really good solution!

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
8 * 2 XP za poprawkę interfejsu w Cars (3522)
10 XP za testy w Cars (3523)
3 * 2 XP za punktualność (1119)
5 XP za wyróżnione rozwiązanie (1120)
{ "requirement_ids": [3521, 3523, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 3522, 1119, 1119, 1119, 1120] }

**Pull request closed without merging**